src/webgpu/web_platform/copyToTexture/canvas.spec.ts

```diff
@@ -286,22 +271,28 @@ g.test('copy_contents_from_2d_context_canvas')
     u
       .combine('canvasType', ['onscreen', 'offscreen'] as const)
       .combine('dstColorFormat', kValidTextureFormatsForCopyIB2T)
-      .combine('premultipliedAlpha', [true, false])
+      .combine('dstPremultiplied', [true, false])
       .beginSubcases()
       .combine('width', [1, 2, 4, 15, 255, 256])
       .combine('height', [1, 2, 4, 15, 255, 256])
   )
   .fn(async t => {
-    const { width, height, canvasType, dstColorFormat, premultipliedAlpha } = t.params;
-
-    // The rgb10a2unorm dst texture will have tiny errors when we compare actual and expectation.
-    // This is due to the convert from 8-bit to 10-bit combined with alpha value ops. So for
-    // rgb10a2unorm dst textures, we'll set alphaValue to 0.0 to test.
+    const { width, height, canvasType, dstColorFormat, dstPremultiplied } = t.params;
+
+    // When dst texture format is rgb10a2unorm, the generated expected value of the result
+    // may have tiny errors compared to the actual result when the channel value is
+    // not 1.0 or 0.0.
+    // For example, we init the pixel with r channel to 0.6. And the denormalized value for
+    // 10-bit channel is 613.8, which needs to call "round" or other function to get an integer.
+    // It is possible that gpu adopt different "round" as our cpu implementation(we use Math.round())
+    // and it will generate tiny errors.
+    // So the cases with rgb10a2unorm dst texture format are handled specially by painting opaque rects
+    // to ensure they will have stable result after alphaOps(should keep the same value).
```

Comment:
Thanks, this makes sense! Does this mean we are losing some coverage, though, as we only test opaque values when the destination is rgb10a2? It's probably not that big of a deal, but perhaps we should leave a TODO to test this with non-opaque values using a tolerance. In theory, this should be much easier at some point later, once we've written good texture checking helpers that can deal with tolerances easily.

Reply:
Yes, I think we lost some coverage and the TODO has already been added in the desc. I'll address it when I figure out the pack(encode) works in framework and how to add the error lost.

Reply:
Ahh, sorry I forgot about the TODO. I had seen it before and forgot :)

Reply:
Also realized that we could do some opt in initialization part to reduce the case running time. Add TODOs for the helper functions.